src/client.js

```diff
@@ -1,52 +1,48 @@
 /* eslint-disable no-var, key-spacing, object-curly-spacing, prefer-arrow-callback, semi, keyword-spacing */
 
-/**
- * @param {number} c Total number of hydration islands
- */
-function initPreactIslands(c) {
-	var el = document.currentScript.parentNode;
-	if (!document.getElementById('praect-island-style')) {
-		var s = document.createElement('style');
-		s.id = 'preact-island-style';
-		s.textContent = 'preact-island{display:contents}';
-		document.head.appendChild(s);
-	}
-	var o = new MutationObserver(function (m) {
-		for (var i = 0; i < m.length; i++) {
-			var added = m[i].addedNodes;
-			for (var j = 0; j < added.length; j++) {
-				if (added[j].nodeType !== 1) continue;
-				var id = added[j].getAttribute('data-target');
-				var target = document.querySelector('[data-id="' + id + '"]');
-				if (target) {
-					while (target.firstChild !== null) {
-						target.removeChild(target.firstChild);
-					}
-					while (added[j].firstChild !== null) {
-						target.appendChild(added[j].firstChild);
-					}
-					target.hydrate = true;
-				}
-				if (--c === 0) {
-					o.disconnect();
-					el.parentNode.removeChild(el);
+function initPreactIslandElement() {
+	class PreactIslandElement extends HTMLElement {
+		connectedCallback() {
+			var d = this;
+			if (!d.isConnected) return;
+
+			let i = this.getAttribute('data-target');
+			if (!i) return;
+
+			var s,
+				e,
+				c = document.createNodeIterator(document, 128);
+			while (c.nextNode()) {
+				let n = c.referenceNode;
+				if (n.data == 'preact-island:' + i) s = n;
+				else if (n.data == '/preact-island:' + i) e = n;
+				if (s && e) break;
+			}
+			if (s && e) {
+				var p = e.previousSibling;
+				while (p != s) {
+					if (!p || p == s) break;
+
+					e.parentNode.removeChild(p);
+					p = e.previousSibling;
 				}
+				while (d.firstChild) e.parentNode.insertBefore(d.firstChild, e);
+
+				d.parentNode.removeChild(d);
 			}
```

Comment:
Doing a deep walk here seems expensive. What about using a hidden element to denote the start of any island, followed by a comment sibling to denote its end?

```html
  <link id="preact-island-00">
  <h1>hello world</h1>
  this is some text
  <!--preact-island-00-->
```

```suggestion
			var n = 'preact-island-' + id,
				s = document.getElementById(n),
				p = s.parentNode,
				e = s,
				b;
			while (e) {
				b = e.nextSibling;
				p.removeChild(e);
				if (e.nodeType === 8 && e.data === n) break;
				e = b;
			}
			while (b = d.firstChild) p.insertBefore(b, e);
```

Reply:
Breaks semantic HTML for lists and whatnot. Same reason to not use a CE for denoting the fallback.

Reply:
Yeah I think there is no way around using comments as we always risk breaking HTML or CSS selectors (`:nth-child`, etc)

Reply:
Does this matter much if the elements are removed immediately?

Reply:
Yes because when the fallback is displayed (potentially a long while) styles / semantics will be broken.